### PR TITLE
METRON-2118: Added a LEEF parser

### PR DESCRIPTION
**Pull request by simonellistonball**
## Contributor Comments
LEEF is a popular format in IBM shops as it is the default supported by Qradar. In a number of ways it is similar to CEF. This PR supports LEEF 1.0 and 2.0 per the IBM guide at https://www.ibm.com/support/knowledgecenter/SS42VS_DSM/c_LEEF_Format_Guide_intro.html and also some found in the wild examples which are technically not up to the IBM spec, and are much closer to the CEF spec.

The CEF parser has been slightly refactored in this effort to expose CEF extension parsing for reuse in the LEEF parser for in the wild examples of CEF style 'delimiters'.

This has been tested against a variety of samples from public sources, and from synthetic data generated according to the spec in new unit tests.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [X] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [X] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [X] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [X] Have you included steps or a guide to how the change may be verified and tested manually?
- [X] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [X] Have you written or updated unit tests and or integration tests to verify your changes?
- [X] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

- [ ] Have you ensured that any documentation diagrams have been updated, along with their source files, using [draw.io](https://www.draw.io/)? See [Metron Development Guidelines](https://cwiki.apache.org/confluence/display/METRON/Development+Guidelines) for instructions.

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

**Comment by ottobackwards** [comment]
Simon this looks very nice, I'm going to review.

One thing I would ask if you can do an integration test, with integration test data and a configuration for this as we have done with the other parsers?

**Comment by mmiklavc** [comment]
Agree with @ottobackwards. I just had to mess about in the integration tests for parsers, so here are some links to help that process:

- https://github.com/apache/metron/tree/master/metron-platform/metron-parsing/metron-parsers-common/src/test/java/org/apache/metron/parsers/integration
- https://github.com/apache/metron/tree/master/metron-platform/metron-integration-test/src/main/sample/data

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+import org.apache.metron.parsers.BasicParser;
+import org.apache.metron.parsers.ParseException;
+import org.apache.metron.parsers.cef.CEFParser;
+import org.apache.metron.parsers.utils.DateUtils;
+import org.apache.metron.parsers.utils.SyslogUtils;
+import org.json.simple.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LEEFParser extends BasicParser {
```
Comment (mmiklavc) [inline metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: Leave the existing files as they are, but for new files can you give these a quick reformat with 2 space indent and tabs as spaces? We're working towards normalizing the file formats, which makes diffs and code reviews easier and faster.

You can manually change it in the IDE:
image

or there's a simple set of instructions for importing the Google stylesheet we use - see section 2.2.2 here - https://cwiki.apache.org/confluence/display/METRON/Development+Guidelines

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+	protected static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+	private static final String HEADER_CAPTURE_PATTERN = "[^\\|]*";
+	private static final Charset UTF_8 = Charset.forName("UTF-8");
+	private static final String EXTENSION_CAPTURE_PATTERN = "(?<!\\\\)=";
+
+	private Pattern p;
+	private Pattern pext;
+
+	public void init() {
```
Comment (mmiklavc) [inline metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: Can you link the LEEF spec you used in a comment?

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+		p = Pattern.compile(pattern);
+
+		pext = Pattern.compile(EXTENSION_CAPTURE_PATTERN);
+	}
+
+	@SuppressWarnings("unchecked")
+	public List<JSONObject> parse(byte[] rawMessage) {
+		List<JSONObject> messages = new ArrayList<>();
+
+		String cefString = new String(rawMessage, UTF_8);
```
Comment (mmiklavc) [inline metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: Just making a note as this relates to another discussion on default charsets - https://github.com/apache/metron/pull/1341#issuecomment-493459303. Regardless of which comes first, we should land on a cohesive strategy for all parsers and be sure this parser has the change once both PRs are complete.

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+				}
+				delimiter = "(?<!\\\\)[" + delimiter.replace("^", "\\^").replace("\t", "\\t") + "]";
+				
+				String[] kvs = ext.split(delimiter); 
+				for (String kv: kvs) {
+					String[] a = kv.split("=");
+					obj.put(a[0], a[1]);
+				}
+			} else if (version.equals("1.0") || version.isEmpty()) {
+				// TODO - Support LEEF 1.0 delimiters
```
Comment (mmiklavc) [inline metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: Is this still "TODO" or is the tab char the delim?

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+				}
+			} else {
+				// Found in the wild examples using CEF rules, which need to handle the processing per the CEFParser
+				// Note that technically LEEF does not support the CEF approach to numbered custom variables.
+				// We however do here, due to some found in the wild exceptions to the standard.
+				CEFParser.parseExtensions(ext, obj);
+			}
+
+			// Rename standard CEF fields to comply with Metron standards
+			obj = mutate(obj, "dst", "ip_dst_addr");
```
Comment (mmiklavc) [inline metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: These fields should reference the Metron constants here - https://github.com/apache/metron/blob/master/metron-platform/metron-common/src/main/java/org/apache/metron/common/Constants.java#L51

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+			obj = mutate(obj, "dstPort", "ip_dst_port");
+			obj = convertToInt(obj, "ip_dst_port");
+
+			obj = mutate(obj, "src", "ip_src_addr");
+			obj = mutate(obj, "srcPort", "ip_src_port");
+			obj = convertToInt(obj, "ip_src_port");
+
+			obj.put("original_string", cefString);
```
Comment (mmiklavc) [inline metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: I have a PR up for review to fix original_string handling - assuming it goes in, this can probably be removed, or at the very least reference the constant field https://github.com/apache/metron/blob/master/metron-platform/metron-common/src/main/java/org/apache/metron/common/Constants.java#L58

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+		return obj;
+	}
+
+	private void headerBlock(String name, StringBuilder sb) {
+		sb.append("(?<").append(name).append(">").append(HEADER_CAPTURE_PATTERN).append(")");
+	}
+
+	@Override
+	public void configure(Map<String, Object> config) {
+		// TODO Auto-generated method stub
```
Comment (mmiklavc) [inline metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: Let's get rid of this comment

File: metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+import com.github.fge.jsonschema.core.report.ProcessingReport;
+import com.github.fge.jsonschema.main.JsonSchemaFactory;
+import com.github.fge.jsonschema.main.JsonValidator;
+import com.google.common.io.Resources;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+public class LEEFParserTest {
+	private static final Charset UTF_8 = Charset.forName("utf-8");
```
Comment (mmiklavc) [inline metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java]: Please use `java.nio.charset.StandardCharsets#UTF_8`

File: metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+			try {
+				json = (Map<?, ?>) parser.parse(parsed.toJSONString());
+				Assert.assertEquals(true, validateJsonData(schema, json.toString()));
+			} catch (ParseException e) {
+				e.printStackTrace();
+			}
+
+			// test against an explicit json example
+			if (!targetJson.isEmpty()) {
+
```
Comment (mmiklavc) [inline metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java]: Are we testing anything here?

File: metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+			Assert.assertTrue((long) parsed.get("timestamp") > 0);
+
+			JSONParser parser = new JSONParser();
+
+			Map<?, ?> json = null;
+			try {
+				json = (Map<?, ?>) parser.parse(parsed.toJSONString());
+				Assert.assertEquals(true, validateJsonData(schema, json.toString()));
+			} catch (ParseException e) {
+				e.printStackTrace();
```
Comment (mmiklavc) [inline metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java]: If the parse fails, it's not going to fail the test with this `try/catch` block, and it's going to miss the `Assert.assertEquals` above. Can you tweak this and please remove the `printStackTrace` as well?

File: metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+			if (!targetJson.isEmpty()) {
+
+			}
+		}
+	}
+
+	private void assertSimpleSample(List<JSONObject> parse) {
+		JSONObject obj = parse.get(0);
+		Assert.assertNotNull(obj);
+		Assert.assertTrue(obj.containsKey("ip_src_addr"));
```
Comment (mmiklavc) [inline metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java]: All common fields should use the keys defined in the Constants file https://github.com/apache/metron/blob/master/metron-platform/metron-common/src/main/java/org/apache/metron/common/Constants.java#L51

File: metron-platform/metron-parsing/metron-parsers/src/test/resources/org/apache/metron/parsers/leef/sample.schema [hunk 12]
```diff
@@ -0,0 +1,27 @@
+{
```
Comment (mmiklavc) [inline metron-platform/metron-parsing/metron-parsers/src/test/resources/org/apache/metron/parsers/leef/sample.schema]: Cool!

File: metron-platform/metron-parsing/metron-parsers-common/src/main/java/org/apache/metron/parsers/utils/DateUtils.java [hunk 13]
```diff
@@ -36,6 +36,14 @@
  */
 public class DateUtils {
```
Comment (ottobackwards) [inline metron-platform/metron-parsing/metron-parsers-common/src/main/java/org/apache/metron/parsers/utils/DateUtils.java]: We mention specs here, we should link to them

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/cef/CEFParser.java [hunk 14]
```diff
@@ -45,7 +45,7 @@
 	private static final Charset UTF_8 = Charset.forName("UTF-8");
 
 	private Pattern p;
```
Comment (ottobackwards) [inline metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/cef/CEFParser.java]: +1 to the static pattern

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+	private static final String EXTENSION_CAPTURE_PATTERN = "(?<!\\\\)=";
+
+	private Pattern p;
+	private Pattern pext;
+
+	public void init() {
+
+		// CEF Headers: Device Vendor|Device Product|Device Version|Device Event
+		// Class ID|Name|Severity
+
```
Comment (ottobackwards) [inline metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: these constant strings can be.... content strings ( static final etc )

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+		String cefString = new String(rawMessage, UTF_8);
+
+		Matcher matcher = p.matcher(cefString);
+
+		while (matcher.find()) {
+			JSONObject obj = new JSONObject();
+			if (!matcher.matches()) {
+				break;
+			}
```
Comment (ottobackwards) [inline metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: A lot of these string can be static final defines

**Comment by simonellistonball** [comment]
Lots of good catches there Mike, I'll do some clean up. Many of the issues are inherited from the fact that I heavily 'borrowed' from the existing CEF parser. Do you think it would be worth fixing that up at the same time on this PR, since I'm refactoring bits of it anyway?

**Comment by simonellistonball** [comment]
@ottobackwards sure, integration tests will come shortly. I've not had much luck finding good sample data to make those exercise much more than the unit tests do already (except of course for the pure integration with the parser runner). I would love to see if anyone else has better samples they could contribute to beef up the tests and help us find any edge cases too.

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+		// extension capture:
+		sb.append(" ?(?<extensions>.*)");
+		String pattern = sb.toString();
+
+		p = Pattern.compile(pattern);
+
+		pext = Pattern.compile(EXTENSION_CAPTURE_PATTERN);
+	}
+
+	@SuppressWarnings("unchecked")
```
Comment (ottobackwards) [inline metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: I don't think new parsers should implement the old interface method, this should instead implement the new interface method 
`Optional<MessageParserResult<T>> parseOptionalResult(byte[] parseMessage)`

Comment (ottobackwards) [reply metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: We could also support multiple messages, but that could be a follow on

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java (hunk 3 above)
Comment (ottobackwards) [inline metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: This is LEEF not cef, can we use variable names that match?

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+import org.slf4j.LoggerFactory;
+
+public class LEEFParser extends BasicParser {
+	private static final long serialVersionUID = 1L;
+
+	protected static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+	private static final String HEADER_CAPTURE_PATTERN = "[^\\|]*";
+	private static final Charset UTF_8 = Charset.forName("UTF-8");
+	private static final String EXTENSION_CAPTURE_PATTERN = "(?<!\\\\)=";
+
```
Comment (ottobackwards) [inline metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: Please use descriptive variable names like extensionPattern and otherPattern

File: metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+public class LEEFParserTest {
+	private static final Charset UTF_8 = Charset.forName("utf-8");
+	private LEEFParser parser;
+
+	@Before
+	public void setUp() {
+		parser = new LEEFParser();
+		parser.init();
+	}
+
```
Comment (ottobackwards) [inline metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java]: Is this the only invalid case?  Are there almost could match cases that will fail or should fail?

**Comment by simonellistonball** [comment]
> Agree with @ottobackwards. I just had to mess about in the integration tests for parsers, so here are some links to help that process:
> 
> * https://github.com/apache/metron/tree/master/metron-platform/metron-parsing/metron-parsers-common/src/test/java/org/apache/metron/parsers/integration
> * https://github.com/apache/metron/tree/master/metron-platform/metron-integration-test/src/main/sample/data

Is it really a good idea to keep integration tests for the metron-parsers package in the metron-parsers-common package? Seems like we're mixing concerns there.

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java (hunk 4 above)
Comment (simonellistonball) [reply metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: Nope. I did it and forgot to remove the comment. Done.

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java (hunk 7 above)
Comment (simonellistonball) [reply metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: done

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java (hunk 6 above)
Comment (simonellistonball) [reply metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: This can probably be removed after your PR, though there is also an argument to follow on from your PR with a general clean-up of all the out of the box parsers, which feels like a separate unit of work to me. Should this stand for consistency right now, and wait for yours to go in then follow on as part of the general cleaning? Otherwise yours will have to be in before this in order for tests to pass.

File: metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java (hunk 8 above)
Comment (simonellistonball) [reply metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java]: Done

File: metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java (hunk 9 above)
Comment (simonellistonball) [reply metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java]: Not really, this is a left over from a half complete idea in the current product's CEF Tests

Comment (simonellistonball) [reply metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java]: I will remove both

File: metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java (hunk 10 above)
Comment (simonellistonball) [reply metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java]: Again, this is an unclean leftover, will remove from both.

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java (hunk 2 above)
Comment (simonellistonball) [reply metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: Sure, I'll add that to java doc for the class as a whole, since it applies across the whole implementation.

File: metron-platform/metron-parsing/metron-parsers-common/src/main/java/org/apache/metron/parsers/utils/DateUtils.java (hunk 13 above)
Comment (simonellistonball) [reply metron-platform/metron-parsing/metron-parsers-common/src/main/java/org/apache/metron/parsers/utils/DateUtils.java]: Done

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java (hunk 3 above)
Comment (simonellistonball) [reply metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: Sure, can you guess where I started from with this class ;)

**Comment by mmiklavc** [comment]
> > Agree with @ottobackwards. I just had to mess about in the integration tests for parsers, so here are some links to help that process:
> > 
> > * https://github.com/apache/metron/tree/master/metron-platform/metron-parsing/metron-parsers-common/src/test/java/org/apache/metron/parsers/integration
> > * https://github.com/apache/metron/tree/master/metron-platform/metron-integration-test/src/main/sample/data
> 
> Is it really a good idea to keep integration tests for the metron-parsers package in the metron-parsers-common package? Seems like we're mixing concerns there.

I tend to agree, though a handful of them are considered "common."

- "jsonMap",
- "jsonMapQuery",
- "jsonMapWrappedQuery",
- "syslog3164",
- "syslog5424"

This should probably be split up a bit, but I don't think it needs to happen in this PR.

**Comment by mmiklavc** [comment]
> Lots of good catches there Mike, I'll do some clean up. Many of the issues are inherited from the fact that I heavily 'borrowed' from the existing CEF parser. Do you think it would be worth fixing that up at the same time on this PR, since I'm refactoring bits of it anyway?

Sure, that sounds reasonable. What did you have in mind?

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java (hunk 6 above)
Comment (mmiklavc) [reply metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: > This can probably be removed after your PR, though there is also an argument to follow on from your PR with a general clean-up of all the out of the box parsers, which feels like a separate unit of work to me. Should this stand for consistency right now, and wait for yours to go in then follow on as part of the general cleaning? Otherwise yours will have to be in before this in order for tests to pass.

That sounds good - re: keep your original_string as is for now (though, do reference the constant fields)

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java (hunk 17 above)
Comment (simonellistonball) [reply metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: Fair enough, this was mainly kept the same to ensure similarity with the CEFParser. I've moved to the new API in this case, but there should be a follow on for converting other parsers.

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java (hunk 18 above)
Comment (simonellistonball) [reply metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: Done, mainly by removing the need for the extension pattern, since it wasn't actually any use.

File: metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java (hunk 19 above)
Comment (simonellistonball) [reply metron-platform/metron-parsing/metron-parsers/src/test/java/org/apache/metron/parsers/leef/LEEFParserTest.java]: I've added another, but there are no doubt more delinquent cases we will find in the wild.

**Comment by ottobackwards** [comment]
I would leave the parsers over all testing refactor ( the configurations required for integration getting installed ) for a separate pr and not mix concerns here.  Let's get a standard parser through in this pr.

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java (hunk 15 above)
Comment (simonellistonball) [reply metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: And done. Technically a good compiler does a lot of this for you with string pooling, and these are not things that will be likely to change, but it does help avoid errors I guess.

**Comment by simonellistonball** [comment]
> > Lots of good catches there Mike, I'll do some clean up. Many of the issues are inherited from the fact that I heavily 'borrowed' from the existing CEF parser. Do you think it would be worth fixing that up at the same time on this PR, since I'm refactoring bits of it anyway?
> 
> Sure, that sounds reasonable. What did you have in mind?

I was thinking that the same string to constant work done here in the LEEF parser should also be backported to the CEF parser, but that's arguably a follow on task from this PR since it's in a different parser.

**Comment by simonellistonball** [comment]
I think I've addressed all the good points raised by @ottobackwards and @mmiklavc Anything else blocking this for anyone?

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/cef/CEFParser.java [hunk 20]
```diff
@@ -45,7 +45,7 @@
...
-	private Pattern pext;
+	private static final Pattern pext = Pattern.compile(EXTENSION_CAPTURE_PATTERN);
```
Comment (ottobackwards) [inline metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/cef/CEFParser.java]: sorry, can we give this a better name too

**Comment by ottobackwards** [comment]
nice work simon, i have one more nit comment, after that I'm set to approve

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/cef/CEFParser.java (hunk 20 above)
Comment (simonellistonball) [reply metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/cef/CEFParser.java]: I kinda left that as out of scope since it's in the CEF parser, not the LEEF parser (the LEEF version doesn't use that technique). I'll fix it anyway, although I would say it was a bit of out of the scope of this work.

Comment (ottobackwards) [reply metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/cef/CEFParser.java]: you are right, my mistake.  But if we have already touched the file, might as well, but if you don't no biggie

**Comment by ottobackwards** [comment]
@mmiklavc you all set with this?

File: metron-platform/metron-parsing/README.md [hunk 21]
```diff
@@ -624,6 +624,8 @@ Java parser adapters are intended for higher-velocity topologies and are not eas
 * org.apache.metron.parsers.lancope.BasicLancopeParser : Parse Lancope messages
 * org.apache.metron.parsers.syslog.Syslog5424Parser : Parse Syslog RFC 5424 messages
 * org.apache.metron.parsers.syslog.Syslog3164Parser : Parse Syslog RFC 3164 messages
+* org.apache.metron.parsers.cef.CEFParser: Parse CEF format messages
```
Comment (mmiklavc) [inline metron-platform/metron-parsing/README.md]: @ottobackwards @simonellistonball - Do we not have any documentation for configuring any of the Java parsers? Not really of consequence for the LEEF parser since the config method is empty, but what about the others? e.g. `BaseSyslogParser` appears to have some config around device timezone, but there's no mention of it in our README.

Comment (simonellistonball) [reply metron-platform/metron-parsing/README.md]: Not really at the moment. We could clean that up, but in this case, LEEF is leaf. As you say, in this case we don't really have anything to document, since there is no config.

**Comment by mmiklavc** [comment]
Looks good, +1 by inspection. Thanks @simonellistonball.

**Comment by ottobackwards** [comment]
+1

File: metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java (hunk 1 above)
Comment (mmiklavc) [reply metron-platform/metron-parsing/metron-parsers/src/main/java/org/apache/metron/parsers/leef/LEEFParser.java]: @simonellistonball - I noticed this after committing, but what happened to fixing format on the new files, e.g. LEEFParser.java, LEEFParserTest.java? They're still 4 space/tab-indented.
